Bug 1848945: OpenStack - Documentation for adding worker nodes using ansible (#4205)
openshift/installer: docs/user/openstack/add_worker_nodes_by_hand.md

#Adding workers nodes by hand
The prefered method of adding worker nodes to an OpenShift cluster is
to use the a worker machineset and machine api operator as described in
https://github.com/openshift/machine-api-operator.

It is also possible to add a worker through other tools like ansible. For example, 
the workers may be added as described in [Installing OpenShift on OpenStack User-Provisioned Infrastructure][1];
or when scaling up the workers using [Openshift-ansible][2]

When doing so, there are a few important points that should be taken into account. 

> ##Apply Necessary Security Groups To Nodes

Author: should be " Apply Necessary Security Groups To Ports"

When adding a new worker node, the necessary security groups need to be applied so the node is 
allowed access the cluster API endpoints and services.   

> If the node being added is attached the cluster os network of an existing cluster,  the pre-existing worker security groups maybe 

Author: "os network" os a pretty obscure term. Probably should use something else more descriptive. os network refers to the openstack network being used to connect the nodes together.

> If the node being added is attached the cluster os network of an existing cluster,  the pre-existing worker security groups maybe 

Author: "being added is attached *to* the os network of an existing cluster"

> applied to the port attaching the node to the cluster os network. Little to no modification needs to 

Author: how many times do we need to say "cluster os network" :)

be made to the existing security group rules.  

> If the nodes being added is attached to a different network than the original os cluster network,

Author: reword. phrase is pretty comboluted.

> then new rules will have to be configured because the rules match on both the destination and the source networks.

Author: reword this.

> ##Configure Allowed Address Pairs on port

Author: caps

The correct address pairs need to be configured on the port that attaches the new node to the os cluster network.  

> The allowed address pairs should configure the port to accept traffic for the three ip addresses of the apiVIP,

Author: ip -> IP

> The allowed address pairs should configure the port to accept traffic for the three ip addresses of the apiVIP,

Author: apiVIP - API VIP, ingressVIP -> ingress VIP

For example: 
```$ openstack port show 97f73ba7-e104-49e2-ad7a-d4b440acc57e -c allowed_address_pairs
   +-----------------------+----------------------------------------------------------+
   | Field                 | Value                                                    |
   +-----------------------+----------------------------------------------------------+
   | allowed_address_pairs | ip_address='10.0.128.5', mac_address='fa:16:3e:0d:35:f8' |
   |                       | ip_address='10.0.128.6', mac_address='fa:16:3e:0d:35:f8' |
   |                       | ip_address='10.0.128.7', mac_address='fa:16:3e:0d:35:f8' |
   +-----------------------+----------------------------------------------------------+
```

*Note: When configuring the allowed address pairs the MAC address should *not* be specified. 
For example: 

```openstack port set <port_id> --allowed-address ip-address='192.168.0.7'```

[1]: https://github.com/openshift/installer/blob/master/docs/user/openstack/install_upi.md#installing-openshift-on-openstack-user-provisioned-infrastructure
[2]: https://github.com/openshift/openshift-ansible

##Access to API Endpoints using hostnames.

> Before any node maybe added to a cluster manually, the node needs to have access to the cluster API endpoints using

Author: manually = using ansible playbook

# Adding Compute Nodes By Hand
The preferred method of adding compute nodes to an OpenShift cluster is
to use the worker MachineSet with [Machine API Operator](https://github.com/openshift/machine-api-operator).

It is also possible to add a compute node through other tools like ansible. For example, 
the compute nodes may be added as described in [Installing OpenShift on OpenStack User-Provisioned Infrastructure][1];
or when scaling up the compute nodes using [Openshift-ansible][2].

When doing so, there are a few important points that should be taken into account. 

## Apply Necessary Security Groups To Ports
When adding a new compute, the necessary security groups need to be applied to 
allow the node access the cluster API endpoints and services.   

If the node is directly attached to the machineNetwork,  the pre-existing worker security groups maybe 
applied to the port attaching the node. The existing security rules should not require modifications.  

If the nodes is attached to a different network than the machineNetwork,

> then new rules will have to be configured because the rules deployed by default on a cluster

Reviewer: What rules?

Author: security group rules. will chance

match on both destination and source networks.

## Configure Allowed Address Pairs On Node Port
The correct address pairs need to be configured on the port attaching the new node to the machineNetwork.  
The allowed address pairs should configure the port to accept traffic for
ingress VIP.  
For example: 

> ```$ openstack port show 97f73ba7-e104-49e2-ad7a-d4b440acc57e -c allowed_address_pairs

Reviewer: You should probably explain how to set it before showing the result of what it should look like.

Author: done

> #Adding workers nodes by hand

Reviewer 2: Add a space between `#` and title to apply format

The prefered method of adding worker nodes to an OpenShift cluster is
to use the a worker machineset and machine api operator as described in
https://github.com/openshift/machine-api-operator.

It is also possible to add a worker through other tools like ansible. For example, 
the workers may be added as described in [Installing OpenShift on OpenStack User-Provisioned Infrastructure][1];
or when scaling up the workers using [Openshift-ansible][2]

When doing so, there are a few important points that should be taken into account. 

##Apply Necessary Security Groups To Ports
When adding a new worker node, the necessary security groups need to be applied so the node is 
allowed access the cluster API endpoints and services.   

If the node is directly attached the machineNetwork,  the pre-existing worker security groups maybe 

> applied to the port attaching the node. The security rules may require little to no modifications.  

Reviewer 3: ```suggestion
applied to the port attaching the node. The security rules should not require modifications.  
```

Just so people arent confused about whether they need to change them.

> If the nodes being added is attached to a different network than the machineNetwork,

Author: I would have thought so as well, but we already had a customer who was able to attach compute nodes to a different network than "machineNetwork" usin UPI, and they had to edit the security groups to correct for this.

then new rules will have to be configured because the rules deployed by default on a cluster

> match on both destination and source networks.

Reviewer 3: How will the security group rules need to be changed in this use case?

Author: Its dificult to predict how the security groups need to be modified because this configuration (different network than machineNetwork), pretty much is everything else. 
not sure we can give more detail on how to do it without leaving out 90% of cases.

##Configure Allowed Address Pairs On Node Port
The correct address pairs need to be configured on the port that attaches the new node to the machineNetwork.  

> The allowed address pairs should configure the port to accept traffic for the three IP addresses of the API VIP,

Author: The example I added here was a copy and paste directly from a compute node that was added by the IPI installation. But as you point out, in reality we only need to do the ingress VIP, should we add that detail? 
how could we phrase that?  and would adding all three IPs be a problem. 
Just trying to explain without having to go into a full blown exaplanation as to why we need to do this.

> The allowed address pairs should configure the port to accept traffic for the three IP addresses of the API VIP,

Reviewer: Right, it doesn't hurt setting the allowed_address_pairs and not using it.

ingress VIP, and cluster DNS service.  
For example: 
```$ openstack port show 97f73ba7-e104-49e2-ad7a-d4b440acc57e -c allowed_address_pairs
   +-----------------------+----------------------------------------------------------+
   | Field                 | Value                                                    |
   +-----------------------+----------------------------------------------------------+
   | allowed_address_pairs | ip_address='10.0.128.5', mac_address='fa:16:3e:0d:35:f8' |
   |                       | ip_address='10.0.128.6', mac_address='fa:16:3e:0d:35:f8' |
   |                       | ip_address='10.0.128.7', mac_address='fa:16:3e:0d:35:f8' |
   +-----------------------+----------------------------------------------------------+
```

*Note: When configuring the allowed address pairs the MAC address should *not* be specified. 
For example: 

> ```openstack port set <port_id> --allowed-address ip-address='192.168.0.7'```

Author: what would be the best way to find out which is the ingressVIP?

> ```openstack port set <port_id> --allowed-address ip-address='192.168.0.7'```

Reviewer: The most reliable way is probably to check for the ingress port configuration:
```
openstack port show <cluster-id>-ingress-port -f value -c fixed_ips
```

> ```openstack port set <port_id> --allowed-address ip-address='192.168.0.7'```

Author: done

[2]: https://github.com/openshift/openshift-ansible

##Access To API Endpoints Using Hostnames.
The node needs to have access to the cluster API endpoints using
fully qualified hostnames.  The node's DNS configuration should resolve the hostname to an IP address which is accessible
to the node. Network security should be configured to allow access.

Before starting the deployment process
access maybe verified by using appropriate tools.  For example, if the cluster is named
my.cluster.com, then the following _curl_ command maybe used to check the necessary access: 
``` curl --insecure https://api.url.text:6443```
where ```api.url.text``` can be any hostname that resolves to an IP address the node can use to reach the API endpoints. 
It could be  _api-int.my.cluster_ or _api.my.cluster_, or _somethingelse.my.cluster_. The actual hostname is irrelevant. What is important is that the node 
translates the hostname to an IP address it can use to access the cluster API endpoints. It could be an internal ip address if the node
 is directly connected to the machineNetwork, or it could be a FIP address which maps to the cluster API endpoint, if the nodes needs to 

>  go through the external network to reach the API endpoints. 

Reviewer: This is confusing. I suggest you pick one example.

then new security group rules will have to be configured because the security group rules deployed by default on a cluster
match on both destination and source networks.

## Configure Allowed Address Pairs On Node Port
The correct address pairs need to be configured on the port attaching the new node to the machineNetwork.  
The allowed address pairs should configure the port to accept traffic for
ingress VIP.  
For example: 

```

$ openstack port set 97f73ba7-e104-49e2-ad7a-d4b440acc57e --allowed-address ip-address='10.0.128.7'
$
$ openstack port show 97f73ba7-e104-49e2-ad7a-d4b440acc57e -c allowed_address_pairs
   +-----------------------+----------------------------------------------------------+
   | Field                 | Value                                                    |
   +-----------------------+----------------------------------------------------------+
   | allowed_address_pairs | ip_address='10.0.128.7', mac_address='fa:16:3e:0d:35:f8' |
   +-----------------------+----------------------------------------------------------+
```

*Note: When configuring the allowed address pairs the MAC address should *not* be specified. 

> For example: 

Reviewer: Not needed, because we already provide the command to set the allowed address pair on the port.

> The correct address pairs need to be configured on the port attaching the new node to the machineNetwork.  
> The allowed address pairs should configure the port to accept traffic for
> ingress VIP.  
> For example: 
> 
> 
> ```
> 
> $ openstack port set 97f73ba7-e104-49e2-ad7a-d4b440acc57e --allowed-address ip-address='10.0.128.7'
> $
> $ openstack port show 97f73ba7-e104-49e2-ad7a-d4b440acc57e -c allowed_address_pairs
>    +-----------------------+----------------------------------------------------------+
>    | Field                 | Value                                                    |
>    +-----------------------+----------------------------------------------------------+
>    | allowed_address_pairs | ip_address='10.0.128.7', mac_address='fa:16:3e:0d:35:f8' |
>    +-----------------------+----------------------------------------------------------+
> ```
> 
> *Note: When configuring the allowed address pairs the MAC address should *not* be specified. 
> For example: 
> 
> The value for the ingress VIP can be found by running the command: 
> 
> ```openstack port show <cluster-id>-ingress-port -f value -c fixed_ips```

Reviewer: The flow is backward. I suggest you change the order:

```suggestion
The port for the new machines needs to be configured to accept traffic for the ingress VIP.

To find the value for the ingress VIP, run the command: 
\```
INGRESS_VIP=$(openstack port show <cluster-id>-ingress-port -f value -c fixed_ips)
\```

Then change the port setting to allow traffic bound for the ingress VIP:
\```
openstack port set <port-id> --allowed-address ip-address=${INGRESS_VIP}
\```
```

Author: done
